// @wix/filter-builder/dist/cjs/type-utils.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.isArray = isArray;
exports.isBoolean = isBoolean;
exports.isDate = isDate;
exports.isFunction = isFunction;
exports.isNull = isNull;
exports.isNumber = isNumber;
exports.isObject = isObject;
exports.isString = isString;
exports.isUndefined = isUndefined;
exports.typeForDisplay = typeForDisplay;
var _kindOf = _interopRequireDefault(require("kind-of"));
function isObject(obj) {
  return (0, _kindOf.default)(obj) === 'object';
}
function isArray(arr) {
  return (0, _kindOf.default)(arr) === 'array';
}
function isDate(obj) {
  return (0, _kindOf.default)(obj) === 'date';
}
function isNumber(obj) {
  return (0, _kindOf.default)(obj) === 'number';
}
function isBoolean(obj) {
  return (0, _kindOf.default)(obj) === 'boolean';
}
function isString(obj) {
  return (0, _kindOf.default)(obj) === 'string';
}
function isFunction(obj) {
  return (0, _kindOf.default)(obj) === 'function';
}
function typeForDisplay(obj) {
  return upperCaseFirst((0, _kindOf.default)(obj));
}
function isUndefined(obj) {
  return (0, _kindOf.default)(obj) === 'undefined';
}
function isNull(obj) {
  return (0, _kindOf.default)(obj) === 'null';
}
function upperCaseFirst(str) {
  if (!isString(str)) {
    return str;
  }
  const first = str.slice(0, 1).toUpperCase();
  const rest = str.slice(1, str.length);
  return first + rest;
}
