// @wix/identity/build/cjs/src/iam-recovery-v1-recovery.public.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Reason = exports.StatusName = exports.AddressTag = exports.PhoneTag = exports.EmailTag = exports.PrivacyStatus = exports.StateType = exports.TenantType = exports.recover = exports.sendActivationEmail = exports.sendRecoveryEmail = exports.__metadata = void 0;
const iam_recovery_v1_recovery_universal_1 = require("./iam-recovery-v1-recovery.universal");
exports.__metadata = { PACKAGE_NAME: '@wix/identity' };
function sendRecoveryEmail(httpClient) {
    return (email, options) => (0, iam_recovery_v1_recovery_universal_1.sendRecoveryEmail)(email, options, 
    // @ts-ignore
    { httpClient });
}
exports.sendRecoveryEmail = sendRecoveryEmail;
function sendActivationEmail(httpClient) {
    return (identityId, options) => (0, iam_recovery_v1_recovery_universal_1.sendActivationEmail)(identityId, options, 
    // @ts-ignore
    { httpClient });
}
exports.sendActivationEmail = sendActivationEmail;
function recover(httpClient) {
    return (recoveryToken, options) => (0, iam_recovery_v1_recovery_universal_1.recover)(recoveryToken, options, 
    // @ts-ignore
    { httpClient });
}
exports.recover = recover;
var iam_recovery_v1_recovery_universal_2 = require("./iam-recovery-v1-recovery.universal");
Object.defineProperty(exports, "TenantType", { enumerable: true, get: function () { return iam_recovery_v1_recovery_universal_2.TenantType; } });
Object.defineProperty(exports, "StateType", { enumerable: true, get: function () { return iam_recovery_v1_recovery_universal_2.StateType; } });
Object.defineProperty(exports, "PrivacyStatus", { enumerable: true, get: function () { return iam_recovery_v1_recovery_universal_2.PrivacyStatus; } });
Object.defineProperty(exports, "EmailTag", { enumerable: true, get: function () { return iam_recovery_v1_recovery_universal_2.EmailTag; } });
Object.defineProperty(exports, "PhoneTag", { enumerable: true, get: function () { return iam_recovery_v1_recovery_universal_2.PhoneTag; } });
Object.defineProperty(exports, "AddressTag", { enumerable: true, get: function () { return iam_recovery_v1_recovery_universal_2.AddressTag; } });
Object.defineProperty(exports, "StatusName", { enumerable: true, get: function () { return iam_recovery_v1_recovery_universal_2.StatusName; } });
Object.defineProperty(exports, "Reason", { enumerable: true, get: function () { return iam_recovery_v1_recovery_universal_2.Reason; } });
