// @wix/identity/build/cjs/src/iam-authentication-v1-authentication.public.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TenantType = exports.StateType = exports.StateStatus = exports.Reason = exports.StatusName = exports.AddressTag = exports.PhoneTag = exports.EmailTag = exports.PrivacyStatus = exports.logout = exports.signOn = exports.proceedToNextState = exports.loginWithIdpConnectionTokens = exports.changePassword = exports.loginV2 = exports.registerV2 = exports.login = exports.register = exports.__metadata = void 0;
const iam_authentication_v1_authentication_universal_1 = require("./iam-authentication-v1-authentication.universal");
exports.__metadata = { PACKAGE_NAME: '@wix/identity' };
function register(httpClient) {
    return (identity, options) => (0, iam_authentication_v1_authentication_universal_1.register)(identity, options, 
    // @ts-ignore
    { httpClient });
}
exports.register = register;
function login(httpClient) {
    return (identifier, options) => (0, iam_authentication_v1_authentication_universal_1.login)(identifier, options, 
    // @ts-ignore
    { httpClient });
}
exports.login = login;
function registerV2(httpClient) {
    return (loginId, options) => (0, iam_authentication_v1_authentication_universal_1.registerV2)(loginId, options, 
    // @ts-ignore
    { httpClient });
}
exports.registerV2 = registerV2;
function loginV2(httpClient) {
    return (loginId, options) => (0, iam_authentication_v1_authentication_universal_1.loginV2)(loginId, options, 
    // @ts-ignore
    { httpClient });
}
exports.loginV2 = loginV2;
function changePassword(httpClient) {
    return (newPassword) => (0, iam_authentication_v1_authentication_universal_1.changePassword)(newPassword, 
    // @ts-ignore
    { httpClient });
}
exports.changePassword = changePassword;
function loginWithIdpConnectionTokens(httpClient) {
    return (options) => (0, iam_authentication_v1_authentication_universal_1.loginWithIdpConnectionTokens)(options, 
    // @ts-ignore
    { httpClient });
}
exports.loginWithIdpConnectionTokens = loginWithIdpConnectionTokens;
function proceedToNextState(httpClient) {
    return (options) => (0, iam_authentication_v1_authentication_universal_1.proceedToNextState)(options, 
    // @ts-ignore
    { httpClient });
}
exports.proceedToNextState = proceedToNextState;
function signOn(httpClient) {
    return (loginId, options) => (0, iam_authentication_v1_authentication_universal_1.signOn)(loginId, options, 
    // @ts-ignore
    { httpClient });
}
exports.signOn = signOn;
function logout(httpClient) {
    return (options) => (0, iam_authentication_v1_authentication_universal_1.logout)(options, 
    // @ts-ignore
    { httpClient });
}
exports.logout = logout;
var iam_authentication_v1_authentication_universal_2 = require("./iam-authentication-v1-authentication.universal");
Object.defineProperty(exports, "PrivacyStatus", { enumerable: true, get: function () { return iam_authentication_v1_authentication_universal_2.PrivacyStatus; } });
Object.defineProperty(exports, "EmailTag", { enumerable: true, get: function () { return iam_authentication_v1_authentication_universal_2.EmailTag; } });
Object.defineProperty(exports, "PhoneTag", { enumerable: true, get: function () { return iam_authentication_v1_authentication_universal_2.PhoneTag; } });
Object.defineProperty(exports, "AddressTag", { enumerable: true, get: function () { return iam_authentication_v1_authentication_universal_2.AddressTag; } });
Object.defineProperty(exports, "StatusName", { enumerable: true, get: function () { return iam_authentication_v1_authentication_universal_2.StatusName; } });
Object.defineProperty(exports, "Reason", { enumerable: true, get: function () { return iam_authentication_v1_authentication_universal_2.Reason; } });
Object.defineProperty(exports, "StateStatus", { enumerable: true, get: function () { return iam_authentication_v1_authentication_universal_2.StateStatus; } });
Object.defineProperty(exports, "StateType", { enumerable: true, get: function () { return iam_authentication_v1_authentication_universal_2.StateType; } });
Object.defineProperty(exports, "TenantType", { enumerable: true, get: function () { return iam_authentication_v1_authentication_universal_2.TenantType; } });
