// @wix/identity/build/cjs/src/iam-authentication-v1-authentication.http.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.logout = exports.signOn = exports.proceedToNextState = exports.loginWithIdpConnectionTokens = exports.changePassword = exports.loginV2 = exports.registerV2 = exports.login = exports.register = void 0;
const metro_runtime_1 = require("@wix/metro-runtime");
const ambassador_1 = require("@wix/metro-runtime/ambassador");
const metro_runtime_2 = require("@wix/metro-runtime");
const _changePasswordRequest = {};
const _commonCustomValue = {
    numValue: 'DOUBLE',
    dateValue: 'google.protobuf.Timestamp',
    listValue: '_commonListValue',
    mapValue: '_commonMapValue',
};
const _commonListValue = { value: '_commonCustomValue' };
const _commonMapValue = { value: 'Map#_commonCustomValue' };
const _customField = { value: '_customValue' };
const _customValue = {
    numValue: 'DOUBLE',
    dateValue: 'google.protobuf.Timestamp',
    listValue: '_listValue',
    mapValue: '_mapValue',
};
const _identity = {
    createdDate: 'google.protobuf.Timestamp',
    updatedDate: 'google.protobuf.Timestamp',
    identityProfile: '_identityProfile',
};
const _identityProfile = { customFields: '_customField' };
const _listValue = { value: '_customValue' };
const _loginRequest = {};
const _loginResponse = {
    identity: '_identity',
    additionalData: 'Map#_customValue',
};
const _loginV2Request = {};
const _loginWithIdpConnectionTokensRequest = {};
const _logoutRequest = {};
const _mapValue = { value: 'Map#_customValue' };
const _proceedToNextStateRequest = {};
const _rawHttpResponse = { body: 'BYTES' };
const _registerRequest = { identity: '_identity' };
const _registerV2Request = { profile: '_identityProfile' };
const _signOnRequest = { profile: '_identityProfile' };
const _signOnResponse = {};
const _stateMachineResponse = {
    identity: '_identity',
    additionalData: 'Map#_commonCustomValue',
};
function resolveWixIamAuthenticationV1AuthenticationServiceUrl(opts) {
    const domainToMappings = {
        _: [
            {
                srcPath: '/_api/iam/authentication',
                destPath: '',
            },
        ],
        'manage._base_domain_': [
            {
                srcPath: '/_api/authentication',
                destPath: '',
            },
        ],
        'users._base_domain_': [
            {
                srcPath: '/authentication',
                destPath: '',
            },
            {
                srcPath: '/iam/wix/google',
                destPath: '/v1/sso/callback/root/0e6a50f5-b523-4e29-990d-f37fa2ffdd69',
            },
        ],
        'dev._base_domain_': [
            {
                srcPath: '/_api/iam/authentication',
                destPath: '',
            },
        ],
        'bo._base_domain_': [
            {
                srcPath: '/_api/iam/authentication',
                destPath: '',
            },
        ],
        'www.wixapis.com': [
            {
                srcPath: '/_api/iam/authentication',
                destPath: '',
            },
        ],
        _base_domain_: [
            {
                srcPath: '/_api/iam/authentication',
                destPath: '',
            },
        ],
        'www._base_domain_': [
            {
                srcPath: '/_api/iam/authentication',
                destPath: '',
            },
        ],
    };
    return (0, metro_runtime_2.resolveUrl)(Object.assign(opts, { domainToMappings }));
}
function register(payload) {
    const { toJSON: toReq, fromJSON: fromReq } = (0, ambassador_1.serializer)(_registerRequest, {
        _customField,
        _customValue,
        _identity,
        _identityProfile,
        _listValue,
        _mapValue,
    });
    const { fromJSON: fromRes } = (0, ambassador_1.serializer)(_loginResponse, {
        _customField,
        _customValue,
        _identity,
        _identityProfile,
        _listValue,
        _mapValue,
    });
    function __register({ host }) {
        const serializedData = toReq(payload);
        const metadata = {
            entityFqdn: 'wix.iam.authentication.v1.authentication',
            method: 'POST',
            methodFqn: 'wix.iam.authentication.v1.AuthenticationService.Register',
            url: resolveWixIamAuthenticationV1AuthenticationServiceUrl({
                protoPath: '/v1/register',
                data: serializedData,
                host,
            }),
            data: serializedData,
            transformResponse: fromRes,
        };
        return metadata;
    }
    __register.fromReq = fromReq;
    return __register;
}
exports.register = register;
function login(payload) {
    const { toJSON: toReq, fromJSON: fromReq } = (0, ambassador_1.serializer)(_loginRequest, {});
    const { fromJSON: fromRes } = (0, ambassador_1.serializer)(_loginResponse, {
        _customField,
        _customValue,
        _identity,
        _identityProfile,
        _listValue,
        _mapValue,
    });
    function __login({ host }) {
        const serializedData = toReq(payload);
        const metadata = {
            entityFqdn: 'wix.iam.authentication.v1.authentication',
            method: 'POST',
            methodFqn: 'wix.iam.authentication.v1.AuthenticationService.Login',
            url: resolveWixIamAuthenticationV1AuthenticationServiceUrl({
                protoPath: '/v1/login',
                data: serializedData,
                host,
            }),
            data: serializedData,
            transformResponse: fromRes,
        };
        return metadata;
    }
    __login.fromReq = fromReq;
    return __login;
}
exports.login = login;
/**
 * Registers a new member.
 *
 * Typically, after a sucessful registration, you generate and use member tokens for the
 * registered member so that subsequent API calls are called as part of a member session.
 *
 * If the email used to register the member already exists as a contact email, the registering
 * member need to verify the email address using a code that is sent to the address.
 */
function registerV2(payload) {
    const { toJSON: toReq, fromJSON: fromReq } = (0, ambassador_1.serializer)(_registerV2Request, {
        _customField,
        _customValue,
        _identityProfile,
        _listValue,
        _mapValue,
    });
    const { fromJSON: fromRes } = (0, ambassador_1.serializer)(_stateMachineResponse, {
        _commonCustomValue,
        _commonListValue,
        _commonMapValue,
        _customField,
        _customValue,
        _identity,
        _identityProfile,
        _listValue,
        _mapValue,
    });
    function __registerV2({ host }) {
        const serializedData = toReq(payload);
        const metadata = {
            entityFqdn: 'wix.iam.authentication.v1.authentication',
            method: 'POST',
            methodFqn: 'wix.iam.authentication.v1.AuthenticationService.RegisterV2',
            url: resolveWixIamAuthenticationV1AuthenticationServiceUrl({
                protoPath: '/v2/register',
                data: serializedData,
                host,
            }),
            data: serializedData,
            transformResponse: fromRes,
        };
        return metadata;
    }
    __registerV2.fromReq = fromReq;
    return __registerV2;
}
exports.registerV2 = registerV2;
/**
 * Logs in an existing user.
 *
 * Typically, after a sucessful login, you generate and use member tokens for the
 * logged-in member so that subsequent API calls are called as part of a member session.
 */
function loginV2(payload) {
    const { toJSON: toReq, fromJSON: fromReq } = (0, ambassador_1.serializer)(_loginV2Request, {});
    const { fromJSON: fromRes } = (0, ambassador_1.serializer)(_stateMachineResponse, {
        _commonCustomValue,
        _commonListValue,
        _commonMapValue,
        _customField,
        _customValue,
        _identity,
        _identityProfile,
        _listValue,
        _mapValue,
    });
    function __loginV2({ host }) {
        const serializedData = toReq(payload);
        const metadata = {
            entityFqdn: 'wix.iam.authentication.v1.authentication',
            method: 'POST',
            methodFqn: 'wix.iam.authentication.v1.AuthenticationService.LoginV2',
            url: resolveWixIamAuthenticationV1AuthenticationServiceUrl({
                protoPath: '/v2/login',
                data: serializedData,
                host,
            }),
            data: serializedData,
            transformResponse: fromRes,
        };
        return metadata;
    }
    __loginV2.fromReq = fromReq;
    return __loginV2;
}
exports.loginV2 = loginV2;
/** Changes the password of a logged in user. */
function changePassword(payload) {
    const { toJSON: toReq, fromJSON: fromReq } = (0, ambassador_1.serializer)(_changePasswordRequest, {});
    const { fromJSON: fromRes } = (0, ambassador_1.serializer)(_changePasswordRequest, {});
    function __changePassword({ host }) {
        const serializedData = toReq(payload);
        const metadata = {
            entityFqdn: 'wix.iam.authentication.v1.authentication',
            method: 'POST',
            methodFqn: 'wix.iam.authentication.v1.AuthenticationService.ChangePassword',
            url: resolveWixIamAuthenticationV1AuthenticationServiceUrl({
                protoPath: '/v2/change-password',
                data: serializedData,
                host,
            }),
            data: serializedData,
            transformResponse: fromRes,
        };
        return metadata;
    }
    __changePassword.fromReq = fromReq;
    return __changePassword;
}
exports.changePassword = changePassword;
function loginWithIdpConnectionTokens(payload) {
    const { toJSON: toReq, fromJSON: fromReq } = (0, ambassador_1.serializer)(_loginWithIdpConnectionTokensRequest, {});
    const { fromJSON: fromRes } = (0, ambassador_1.serializer)(_stateMachineResponse, {
        _commonCustomValue,
        _commonListValue,
        _commonMapValue,
        _customField,
        _customValue,
        _identity,
        _identityProfile,
        _listValue,
        _mapValue,
    });
    function __loginWithIdpConnectionTokens({ host }) {
        const serializedData = toReq(payload);
        const metadata = {
            entityFqdn: 'wix.iam.authentication.v1.authentication',
            method: 'POST',
            methodFqn: 'wix.iam.authentication.v1.AuthenticationService.LoginWithIdpConnectionTokens',
            url: resolveWixIamAuthenticationV1AuthenticationServiceUrl({
                protoPath: '/v1/sso/login-with-tokens',
                data: serializedData,
                host,
            }),
            data: serializedData,
            transformResponse: fromRes,
        };
        return metadata;
    }
    __loginWithIdpConnectionTokens.fromReq = fromReq;
    return __loginWithIdpConnectionTokens;
}
exports.loginWithIdpConnectionTokens = loginWithIdpConnectionTokens;
function proceedToNextState(payload) {
    const { toJSON: toReq, fromJSON: fromReq } = (0, ambassador_1.serializer)(_proceedToNextStateRequest, {});
    const { fromJSON: fromRes } = (0, ambassador_1.serializer)(_loginResponse, {
        _customField,
        _customValue,
        _identity,
        _identityProfile,
        _listValue,
        _mapValue,
    });
    function __proceedToNextState({ host }) {
        const serializedData = toReq(payload);
        const metadata = {
            entityFqdn: 'wix.iam.authentication.v1.authentication',
            method: 'POST',
            methodFqn: 'wix.iam.authentication.v1.AuthenticationService.ProceedToNextState',
            url: resolveWixIamAuthenticationV1AuthenticationServiceUrl({
                protoPath: '/v1/state/proceed',
                data: serializedData,
                host,
            }),
            data: serializedData,
            transformResponse: fromRes,
        };
        return metadata;
    }
    __proceedToNextState.fromReq = fromReq;
    return __proceedToNextState;
}
exports.proceedToNextState = proceedToNextState;
function signOn(payload) {
    const { toJSON: toReq, fromJSON: fromReq } = (0, ambassador_1.serializer)(_signOnRequest, {
        _customField,
        _customValue,
        _identityProfile,
        _listValue,
        _mapValue,
    });
    const { fromJSON: fromRes } = (0, ambassador_1.serializer)(_signOnResponse, {});
    function __signOn({ host }) {
        const serializedData = toReq(payload);
        const metadata = {
            entityFqdn: 'wix.iam.authentication.v1.authentication',
            method: 'POST',
            methodFqn: 'wix.iam.authentication.v1.AuthenticationService.SignOn',
            url: resolveWixIamAuthenticationV1AuthenticationServiceUrl({
                protoPath: '/v2/sign-on',
                data: serializedData,
                host,
            }),
            data: serializedData,
            transformResponse: fromRes,
        };
        return metadata;
    }
    __signOn.fromReq = fromReq;
    return __signOn;
}
exports.signOn = signOn;
/** Logs out a member. */
function logout(payload) {
    const { toJSON: toReq, fromJSON: fromReq } = (0, ambassador_1.serializer)(_logoutRequest, {});
    const { fromJSON: fromRes } = (0, ambassador_1.serializer)(_rawHttpResponse, {});
    function __logout({ host }) {
        const serializedData = toReq(payload);
        const metadata = {
            entityFqdn: 'wix.iam.authentication.v1.authentication',
            method: 'GET',
            methodFqn: 'wix.iam.authentication.v1.AuthenticationService.Logout',
            url: resolveWixIamAuthenticationV1AuthenticationServiceUrl({
                protoPath: '/v1/logout',
                data: serializedData,
                host,
            }),
            params: (0, metro_runtime_1.toURLSearchParams)(serializedData),
            transformResponse: fromRes,
        };
        return metadata;
    }
    __logout.fromReq = fromReq;
    return __logout;
}
exports.logout = logout;
