// @wix/identity/build/cjs/src/iam-verification-v1-start-response.public.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Reason = exports.StatusName = exports.AddressTag = exports.PhoneTag = exports.EmailTag = exports.PrivacyStatus = exports.StateType = exports.Target = exports.verifyDuringAuthentication = exports.start = exports.__metadata = void 0;
const iam_verification_v1_start_response_universal_1 = require("./iam-verification-v1-start-response.universal");
exports.__metadata = { PACKAGE_NAME: '@wix/identity' };
function start(httpClient) {
    return (options) => (0, iam_verification_v1_start_response_universal_1.start)(options, 
    // @ts-ignore
    { httpClient });
}
exports.start = start;
function verifyDuringAuthentication(httpClient) {
    return (code, options) => (0, iam_verification_v1_start_response_universal_1.verifyDuringAuthentication)(code, options, 
    // @ts-ignore
    { httpClient });
}
exports.verifyDuringAuthentication = verifyDuringAuthentication;
var iam_verification_v1_start_response_universal_2 = require("./iam-verification-v1-start-response.universal");
Object.defineProperty(exports, "Target", { enumerable: true, get: function () { return iam_verification_v1_start_response_universal_2.Target; } });
Object.defineProperty(exports, "StateType", { enumerable: true, get: function () { return iam_verification_v1_start_response_universal_2.StateType; } });
Object.defineProperty(exports, "PrivacyStatus", { enumerable: true, get: function () { return iam_verification_v1_start_response_universal_2.PrivacyStatus; } });
Object.defineProperty(exports, "EmailTag", { enumerable: true, get: function () { return iam_verification_v1_start_response_universal_2.EmailTag; } });
Object.defineProperty(exports, "PhoneTag", { enumerable: true, get: function () { return iam_verification_v1_start_response_universal_2.PhoneTag; } });
Object.defineProperty(exports, "AddressTag", { enumerable: true, get: function () { return iam_verification_v1_start_response_universal_2.AddressTag; } });
Object.defineProperty(exports, "StatusName", { enumerable: true, get: function () { return iam_verification_v1_start_response_universal_2.StatusName; } });
Object.defineProperty(exports, "Reason", { enumerable: true, get: function () { return iam_verification_v1_start_response_universal_2.Reason; } });
