// @wix/media/build/cjs/src/media-site-media-v1-folder.public.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RootFolder = exports.SortOrder = exports.State = exports.listDeletedFolders = exports.bulkRestoreFoldersFromTrashBin = exports.bulkDeleteFolders = exports.generateFolderDownloadUrl = exports.updateFolder = exports.searchFolders = exports.listFolders = exports.getFolder = exports.createFolder = exports.__metadata = void 0;
const media_site_media_v1_folder_universal_1 = require("./media-site-media-v1-folder.universal");
exports.__metadata = { PACKAGE_NAME: '@wix/media' };
function createFolder(httpClient) {
    return (displayName, options) => (0, media_site_media_v1_folder_universal_1.createFolder)(displayName, options, 
    // @ts-ignore
    { httpClient });
}
exports.createFolder = createFolder;
function getFolder(httpClient) {
    return (folderId) => (0, media_site_media_v1_folder_universal_1.getFolder)(folderId, 
    // @ts-ignore
    { httpClient });
}
exports.getFolder = getFolder;
function listFolders(httpClient) {
    return (options) => (0, media_site_media_v1_folder_universal_1.listFolders)(options, 
    // @ts-ignore
    { httpClient });
}
exports.listFolders = listFolders;
function searchFolders(httpClient) {
    return (options) => (0, media_site_media_v1_folder_universal_1.searchFolders)(options, 
    // @ts-ignore
    { httpClient });
}
exports.searchFolders = searchFolders;
function updateFolder(httpClient) {
    return (_id, folder) => (0, media_site_media_v1_folder_universal_1.updateFolder)(_id, folder, 
    // @ts-ignore
    { httpClient });
}
exports.updateFolder = updateFolder;
function generateFolderDownloadUrl(httpClient) {
    return (folderId) => (0, media_site_media_v1_folder_universal_1.generateFolderDownloadUrl)(folderId, 
    // @ts-ignore
    { httpClient });
}
exports.generateFolderDownloadUrl = generateFolderDownloadUrl;
function bulkDeleteFolders(httpClient) {
    return (folderIds, options) => (0, media_site_media_v1_folder_universal_1.bulkDeleteFolders)(folderIds, options, 
    // @ts-ignore
    { httpClient });
}
exports.bulkDeleteFolders = bulkDeleteFolders;
function bulkRestoreFoldersFromTrashBin(httpClient) {
    return (folderIds) => (0, media_site_media_v1_folder_universal_1.bulkRestoreFoldersFromTrashBin)(folderIds, 
    // @ts-ignore
    { httpClient });
}
exports.bulkRestoreFoldersFromTrashBin = bulkRestoreFoldersFromTrashBin;
function listDeletedFolders(httpClient) {
    return (options) => (0, media_site_media_v1_folder_universal_1.listDeletedFolders)(options, 
    // @ts-ignore
    { httpClient });
}
exports.listDeletedFolders = listDeletedFolders;
var media_site_media_v1_folder_universal_2 = require("./media-site-media-v1-folder.universal");
Object.defineProperty(exports, "State", { enumerable: true, get: function () { return media_site_media_v1_folder_universal_2.State; } });
Object.defineProperty(exports, "SortOrder", { enumerable: true, get: function () { return media_site_media_v1_folder_universal_2.SortOrder; } });
Object.defineProperty(exports, "RootFolder", { enumerable: true, get: function () { return media_site_media_v1_folder_universal_2.RootFolder; } });
