// @wix/media/build/cjs/src/media-site-media-v1-file-descriptor.types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StreamFormat = exports.RootFolder = exports.SortOrder = exports.UploadProtocol = exports.State = exports.OperationStatus = exports.MediaType = void 0;
var MediaType;
(function (MediaType) {
    MediaType["UNKNOWN"] = "UNKNOWN";
    MediaType["IMAGE"] = "IMAGE";
    MediaType["VIDEO"] = "VIDEO";
    MediaType["AUDIO"] = "AUDIO";
    MediaType["DOCUMENT"] = "DOCUMENT";
    MediaType["VECTOR"] = "VECTOR";
    MediaType["ARCHIVE"] = "ARCHIVE";
    MediaType["MODEL3D"] = "MODEL3D";
    MediaType["OTHER"] = "OTHER";
})(MediaType = exports.MediaType || (exports.MediaType = {}));
var OperationStatus;
(function (OperationStatus) {
    /** File upload or processing failed */
    OperationStatus["FAILED"] = "FAILED";
    /** File is ready for consumption */
    OperationStatus["READY"] = "READY";
    /** File is waiting for processing or currently being processed */
    OperationStatus["PENDING"] = "PENDING";
})(OperationStatus = exports.OperationStatus || (exports.OperationStatus = {}));
var State;
(function (State) {
    State["OK"] = "OK";
    State["DELETED"] = "DELETED";
})(State = exports.State || (exports.State = {}));
var UploadProtocol;
(function (UploadProtocol) {
    /** The upload protocol to use for implementing the resumable upload. */
    UploadProtocol["TUS"] = "TUS";
})(UploadProtocol = exports.UploadProtocol || (exports.UploadProtocol = {}));
var SortOrder;
(function (SortOrder) {
    SortOrder["ASC"] = "ASC";
    SortOrder["DESC"] = "DESC";
})(SortOrder = exports.SortOrder || (exports.SortOrder = {}));
var RootFolder;
(function (RootFolder) {
    /** Root of all site media */
    RootFolder["MEDIA_ROOT"] = "MEDIA_ROOT";
    /** Root of the trash system folder */
    RootFolder["TRASH_ROOT"] = "TRASH_ROOT";
    /** Root of all visitor uploads */
    RootFolder["VISITOR_UPLOADS_ROOT"] = "VISITOR_UPLOADS_ROOT";
})(RootFolder = exports.RootFolder || (exports.RootFolder = {}));
var StreamFormat;
(function (StreamFormat) {
    StreamFormat["UNKNOWN"] = "UNKNOWN";
    StreamFormat["HLS"] = "HLS";
    StreamFormat["DASH"] = "DASH";
})(StreamFormat = exports.StreamFormat || (exports.StreamFormat = {}));
