// @wix/metro-runtime/dist/cjs/serializer/velo-converters/address.js
"use strict";

exports.__esModule = true;
exports.address = void 0;
var _lodash = require("lodash");
var _domain = require("../domain");
const sharedFieldNames = ['city', 'subdivision', 'country', 'postalCode'];
const sharedStreetAddressFieldNames = ['name', 'number', 'apt'];
const address = {
  types: ['wix.common.Address'],
  [_domain.ConverterType.FROM_JSON]: {
    /**
     * currently wix.common.Address represents the deserialized object, so we're good.
     * in case it contains some serializable stuff (e.g. int64), we'll have to make some
     * modifications in the type.
     */
    transform: val => {
      return val && (0, _lodash.omitBy)({
        formatted: val.formattedAddress,
        location: val.geocode,
        addressLine1: val.addressLine,
        addressLine2: val.addressLine2,
        streetAddress: val.streetAddress && (0, _lodash.pick)(val.streetAddress, sharedStreetAddressFieldNames),
        ...(0, _lodash.pick)(val, sharedFieldNames)
      }, _lodash.isUndefined);
    }
  },
  [_domain.ConverterType.TO_JSON]: {
    transform: val => {
      return val && (0, _lodash.omitBy)({
        ...(0, _lodash.pick)(val, sharedFieldNames),
        formattedAddress: val.formatted,
        geocode: val.location,
        addressLine: val.addressLine1,
        addressLine2: val.addressLine2,
        streetAddress: val.streetAddress && (0, _lodash.pick)(val.streetAddress, sharedStreetAddressFieldNames)
      }, _lodash.isUndefined);
    }
  }
};
exports.address = address;
