// @wix/metro-runtime/dist/cjs/serializer/velo-converters/video-v2.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.videoV2 = void 0;
var _querystring = _interopRequireDefault(require("querystring"));
var _domain = require("../domain");
var _convertersUtils = require("./converters-utils");
const WIX_VIDEO = 'video';
const videoV2 = {
  types: ['wix.common.VideoV2'],
  [_domain.ConverterType.FROM_JSON]: {
    transform: val => {
      var _val$posters;
      if (!val) {
        return;
      }
      let fileName = '';
      if (val != null && val.filename) {
        fileName = `/${(0, _convertersUtils.encodeText)(val.filename)}`;
      }
      let posterData = '';
      if ((_val$posters = val.posters) != null && _val$posters.length) {
        const _val$posters2 = val.posters,
          poster1 = _val$posters2[0],
          poster2 = _val$posters2[1];
        const poster = poster2 || poster1;
        let posterId = poster.id || '';
        if (!posterId && poster.url) {
          const idx = poster.url.lastIndexOf('/');
          if (idx !== -1) {
            posterId = poster.url.substring(idx + 1);
          }
        }
        if (posterId) {
          posterData = `${_convertersUtils.URL_HASH_PREFIX}posterUri=${posterId}&posterWidth=${poster.width}&posterHeight=${poster.height}`;
        }
      }
      return val.id ? `wix:video://v1/${val.id}${fileName}${posterData}` : val.url;
    }
  },
  [_domain.ConverterType.TO_JSON]: {
    transform: val => {
      if (!val) {
        return;
      }
      const alignedVideo = (0, _convertersUtils.alignIfLegacy)(val, WIX_VIDEO);
      const _URL = new URL(alignedVideo),
        protocol = _URL.protocol,
        hash = _URL.hash,
        pathname = _URL.pathname;
      const _querystring$parse = _querystring.default.parse(hash.replace(_convertersUtils.URL_HASH_PREFIX, '')),
        height = _querystring$parse.posterHeight,
        width = _querystring$parse.posterWidth,
        posterUri = _querystring$parse.posterUri;
      const _pathname$replace$spl = pathname.replace(`${WIX_VIDEO}://v1/`, '').split('/'),
        id = _pathname$replace$spl[0],
        fileName = _pathname$replace$spl[1];
      if (protocol === _convertersUtils.WIX_PROTOCOL) {
        let res = {
          id
        };
        if (fileName) {
          res = {
            ...res,
            filename: (0, _convertersUtils.decodeText)(fileName)
          };
        }
        if (!posterUri) {
          return res;
        }
        return {
          ...res,
          posters: [{
            id: posterUri.toString(),
            height: Number(height),
            width: Number(width)
          }]
        };
      }
      return {
        url: val
      };
    }
  }
};
exports.videoV2 = videoV2;
