// @wix/motion-edm-autogen-transformations-core/dist/cjs/lib/corvid-entity/index.js
"use strict";

exports.__esModule = true;
exports.toCorvidName = exports.isReservedCorvidFieldName = exports.isReservedCommonFieldName = exports.fromCorvidName = exports.TO_CORVID_ENTITY_RENAMED_KEYS = exports.FROM_CORVID_ENTITY_RENAMED_KEYS = void 0;
var _lodash = require("lodash");
const TO_CORVID_ENTITY_RENAMED_KEYS = {
  id: '_id',
  createdDate: '_createdDate',
  updatedDate: '_updatedDate'
};
exports.TO_CORVID_ENTITY_RENAMED_KEYS = TO_CORVID_ENTITY_RENAMED_KEYS;
const FROM_CORVID_ENTITY_RENAMED_KEYS = (0, _lodash.invert)(TO_CORVID_ENTITY_RENAMED_KEYS);
exports.FROM_CORVID_ENTITY_RENAMED_KEYS = FROM_CORVID_ENTITY_RENAMED_KEYS;
const isReservedCorvidFieldName = name => name in FROM_CORVID_ENTITY_RENAMED_KEYS;
exports.isReservedCorvidFieldName = isReservedCorvidFieldName;
const isReservedCommonFieldName = name => name in TO_CORVID_ENTITY_RENAMED_KEYS;
exports.isReservedCommonFieldName = isReservedCommonFieldName;
const toCorvidName = name => TO_CORVID_ENTITY_RENAMED_KEYS[name] || name;
exports.toCorvidName = toCorvidName;
const fromCorvidName = name => FROM_CORVID_ENTITY_RENAMED_KEYS[name] || name;
exports.fromCorvidName = fromCorvidName;
