// @wix/media/build/es/src/media-site-media-v1-folder.meta.js
import * as ambassadorWixMediaSiteMediaV1Folder from './media-site-media-v1-folder.http';
export function createFolder() {
    const payload = {};
    const getRequestOptions = ambassadorWixMediaSiteMediaV1Folder.createFolder(payload);
    const getUrl = (context) => {
        const { url } = getRequestOptions(context);
        return url;
    };
    return {
        getUrl,
        httpMethod: 'POST',
        path: '/v1/folders',
        pathParams: {},
        __requestType: null,
        __originalRequestType: null,
        __responseType: null,
        __originalResponseType: null,
    };
}
export function getFolder() {
    const payload = { folderId: ':folderId' };
    const getRequestOptions = ambassadorWixMediaSiteMediaV1Folder.getFolder(payload);
    const getUrl = (context) => {
        const { url } = getRequestOptions(context);
        return url;
    };
    return {
        getUrl,
        httpMethod: 'GET',
        path: '/v1/folders/{folderId}',
        pathParams: { folderId: 'folderId' },
        __requestType: null,
        __originalRequestType: null,
        __responseType: null,
        __originalResponseType: null,
    };
}
export function listFolders() {
    const payload = {};
    const getRequestOptions = ambassadorWixMediaSiteMediaV1Folder.listFolders(payload);
    const getUrl = (context) => {
        const { url } = getRequestOptions(context);
        return url;
    };
    return {
        getUrl,
        httpMethod: 'GET',
        path: '/v1/folders',
        pathParams: {},
        __requestType: null,
        __originalRequestType: null,
        __responseType: null,
        __originalResponseType: null,
    };
}
export function searchFolders() {
    const payload = {};
    const getRequestOptions = ambassadorWixMediaSiteMediaV1Folder.searchFolders(payload);
    const getUrl = (context) => {
        const { url } = getRequestOptions(context);
        return url;
    };
    return {
        getUrl,
        httpMethod: 'POST',
        path: '/v1/folders/search',
        pathParams: {},
        __requestType: null,
        __originalRequestType: null,
        __responseType: null,
        __originalResponseType: null,
    };
}
export function updateFolder() {
    const payload = { folder: { id: ':folderId' } };
    const getRequestOptions = ambassadorWixMediaSiteMediaV1Folder.updateFolder(payload);
    const getUrl = (context) => {
        const { url } = getRequestOptions(context);
        return url;
    };
    return {
        getUrl,
        httpMethod: 'PATCH',
        path: '/v1/folders/{folder.id}',
        pathParams: { folderId: 'folderId' },
        __requestType: null,
        __originalRequestType: null,
        __responseType: null,
        __originalResponseType: null,
    };
}
export function generateFolderDownloadUrl() {
    const payload = { folderId: ':folderId' };
    const getRequestOptions = ambassadorWixMediaSiteMediaV1Folder.generateFolderDownloadUrl(payload);
    const getUrl = (context) => {
        const { url } = getRequestOptions(context);
        return url;
    };
    return {
        getUrl,
        httpMethod: 'POST',
        path: '/v1/folders/{folderId}/generate-download-url',
        pathParams: { folderId: 'folderId' },
        __requestType: null,
        __originalRequestType: null,
        __responseType: null,
        __originalResponseType: null,
    };
}
export function bulkDeleteFolders() {
    const payload = {};
    const getRequestOptions = ambassadorWixMediaSiteMediaV1Folder.bulkDeleteFolders(payload);
    const getUrl = (context) => {
        const { url } = getRequestOptions(context);
        return url;
    };
    return {
        getUrl,
        httpMethod: 'POST',
        path: '/v1/bulk/folders/delete',
        pathParams: {},
        __requestType: null,
        __originalRequestType: null,
        __responseType: null,
        __originalResponseType: null,
    };
}
export function bulkRestoreFoldersFromTrashBin() {
    const payload = {};
    const getRequestOptions = ambassadorWixMediaSiteMediaV1Folder.bulkRestoreFoldersFromTrashBin(payload);
    const getUrl = (context) => {
        const { url } = getRequestOptions(context);
        return url;
    };
    return {
        getUrl,
        httpMethod: 'POST',
        path: '/v1/bulk/trash-bin/folders/restore',
        pathParams: {},
        __requestType: null,
        __originalRequestType: null,
        __responseType: null,
        __originalResponseType: null,
    };
}
export function listDeletedFolders() {
    const payload = {};
    const getRequestOptions = ambassadorWixMediaSiteMediaV1Folder.listDeletedFolders(payload);
    const getUrl = (context) => {
        const { url } = getRequestOptions(context);
        return url;
    };
    return {
        getUrl,
        httpMethod: 'GET',
        path: '/v1/trash-bin/folders',
        pathParams: {},
        __requestType: null,
        __originalRequestType: null,
        __responseType: null,
        __originalResponseType: null,
    };
}
