// @wix/media/build/es/src/media-site-media-v1-folder.public.js
import { createFolder as universalCreateFolder, getFolder as universalGetFolder, listFolders as universalListFolders, searchFolders as universalSearchFolders, updateFolder as universalUpdateFolder, generateFolderDownloadUrl as universalGenerateFolderDownloadUrl, bulkDeleteFolders as universalBulkDeleteFolders, bulkRestoreFoldersFromTrashBin as universalBulkRestoreFoldersFromTrashBin, listDeletedFolders as universalListDeletedFolders, } from './media-site-media-v1-folder.universal';
export const __metadata = { PACKAGE_NAME: '@wix/media' };
export function createFolder(httpClient) {
    return (displayName, options) => universalCreateFolder(displayName, options, 
    // @ts-ignore
    { httpClient });
}
export function getFolder(httpClient) {
    return (folderId) => universalGetFolder(folderId, 
    // @ts-ignore
    { httpClient });
}
export function listFolders(httpClient) {
    return (options) => universalListFolders(options, 
    // @ts-ignore
    { httpClient });
}
export function searchFolders(httpClient) {
    return (options) => universalSearchFolders(options, 
    // @ts-ignore
    { httpClient });
}
export function updateFolder(httpClient) {
    return (_id, folder) => universalUpdateFolder(_id, folder, 
    // @ts-ignore
    { httpClient });
}
export function generateFolderDownloadUrl(httpClient) {
    return (folderId) => universalGenerateFolderDownloadUrl(folderId, 
    // @ts-ignore
    { httpClient });
}
export function bulkDeleteFolders(httpClient) {
    return (folderIds, options) => universalBulkDeleteFolders(folderIds, options, 
    // @ts-ignore
    { httpClient });
}
export function bulkRestoreFoldersFromTrashBin(httpClient) {
    return (folderIds) => universalBulkRestoreFoldersFromTrashBin(folderIds, 
    // @ts-ignore
    { httpClient });
}
export function listDeletedFolders(httpClient) {
    return (options) => universalListDeletedFolders(options, 
    // @ts-ignore
    { httpClient });
}
export { State, SortOrder, RootFolder, } from './media-site-media-v1-folder.universal';
