// @wix/media/build/es/src/media-site-media-v1-file-descriptor.meta.js
import * as ambassadorWixMediaSiteMediaV1FileDescriptor from './media-site-media-v1-file-descriptor.http';
export function generateFilesDownloadUrl() {
    const payload = {};
    const getRequestOptions = ambassadorWixMediaSiteMediaV1FileDescriptor.generateFilesDownloadUrl(payload);
    const getUrl = (context) => {
        const { url } = getRequestOptions(context);
        return url;
    };
    return {
        getUrl,
        httpMethod: 'POST',
        path: '/v1/files/generate-download-url',
        pathParams: {},
        __requestType: null,
        __originalRequestType: null,
        __responseType: null,
        __originalResponseType: null,
    };
}
export function generateFileDownloadUrl() {
    const payload = { fileId: ':fileId' };
    const getRequestOptions = ambassadorWixMediaSiteMediaV1FileDescriptor.generateFileDownloadUrl(payload);
    const getUrl = (context) => {
        const { url } = getRequestOptions(context);
        return url;
    };
    return {
        getUrl,
        httpMethod: 'POST',
        path: '/v1/files/{fileId}/generate-download-url',
        pathParams: { fileId: 'fileId' },
        __requestType: null,
        __originalRequestType: null,
        __responseType: null,
        __originalResponseType: null,
    };
}
export function getFileDescriptor() {
    const payload = { fileId: ':fileId' };
    const getRequestOptions = ambassadorWixMediaSiteMediaV1FileDescriptor.getFileDescriptor(payload);
    const getUrl = (context) => {
        const { url } = getRequestOptions(context);
        return url;
    };
    return {
        getUrl,
        httpMethod: 'GET',
        path: '/v1/files/{fileId}',
        pathParams: { fileId: 'fileId' },
        __requestType: null,
        __originalRequestType: null,
        __responseType: null,
        __originalResponseType: null,
    };
}
export function getFileDescriptors() {
    const payload = {};
    const getRequestOptions = ambassadorWixMediaSiteMediaV1FileDescriptor.getFileDescriptors(payload);
    const getUrl = (context) => {
        const { url } = getRequestOptions(context);
        return url;
    };
    return {
        getUrl,
        httpMethod: 'POST',
        path: '/v1/files/get-files',
        pathParams: {},
        __requestType: null,
        __originalRequestType: null,
        __responseType: null,
        __originalResponseType: null,
    };
}
export function updateFileDescriptor() {
    const payload = { file: { id: ':fileId' } };
    const getRequestOptions = ambassadorWixMediaSiteMediaV1FileDescriptor.updateFileDescriptor(payload);
    const getUrl = (context) => {
        const { url } = getRequestOptions(context);
        return url;
    };
    return {
        getUrl,
        httpMethod: 'PATCH',
        path: '/v1/files/{file.id}/update',
        pathParams: { fileId: 'fileId' },
        __requestType: null,
        __originalRequestType: null,
        __responseType: null,
        __originalResponseType: null,
    };
}
export function generateFileUploadUrl() {
    const payload = {};
    const getRequestOptions = ambassadorWixMediaSiteMediaV1FileDescriptor.generateFileUploadUrl(payload);
    const getUrl = (context) => {
        const { url } = getRequestOptions(context);
        return url;
    };
    return {
        getUrl,
        httpMethod: 'POST',
        path: '/v1/files/generate-upload-url',
        pathParams: {},
        __requestType: null,
        __originalRequestType: null,
        __responseType: null,
        __originalResponseType: null,
    };
}
export function generateFileResumableUploadUrl() {
    const payload = {};
    const getRequestOptions = ambassadorWixMediaSiteMediaV1FileDescriptor.generateFileResumableUploadUrl(payload);
    const getUrl = (context) => {
        const { url } = getRequestOptions(context);
        return url;
    };
    return {
        getUrl,
        httpMethod: 'POST',
        path: '/v1/files/generate-resumable-upload-url',
        pathParams: {},
        __requestType: null,
        __originalRequestType: null,
        __responseType: null,
        __originalResponseType: null,
    };
}
export function importFile() {
    const payload = {};
    const getRequestOptions = ambassadorWixMediaSiteMediaV1FileDescriptor.importFile(payload);
    const getUrl = (context) => {
        const { url } = getRequestOptions(context);
        return url;
    };
    return {
        getUrl,
        httpMethod: 'POST',
        path: '/v1/files/import',
        pathParams: {},
        __requestType: null,
        __originalRequestType: null,
        __responseType: null,
        __originalResponseType: null,
    };
}
export function bulkImportFiles() {
    const payload = {};
    const getRequestOptions = ambassadorWixMediaSiteMediaV1FileDescriptor.bulkImportFiles(payload);
    const getUrl = (context) => {
        const { url } = getRequestOptions(context);
        return url;
    };
    return {
        getUrl,
        httpMethod: 'POST',
        path: '/v1/bulk/files/import',
        pathParams: {},
        __requestType: null,
        __originalRequestType: null,
        __responseType: null,
        __originalResponseType: null,
    };
}
export function listFiles() {
    const payload = {};
    const getRequestOptions = ambassadorWixMediaSiteMediaV1FileDescriptor.listFiles(payload);
    const getUrl = (context) => {
        const { url } = getRequestOptions(context);
        return url;
    };
    return {
        getUrl,
        httpMethod: 'GET',
        path: '/v1/files',
        pathParams: {},
        __requestType: null,
        __originalRequestType: null,
        __responseType: null,
        __originalResponseType: null,
    };
}
export function searchFiles() {
    const payload = {};
    const getRequestOptions = ambassadorWixMediaSiteMediaV1FileDescriptor.searchFiles(payload);
    const getUrl = (context) => {
        const { url } = getRequestOptions(context);
        return url;
    };
    return {
        getUrl,
        httpMethod: 'POST',
        path: '/v1/files/search',
        pathParams: {},
        __requestType: null,
        __originalRequestType: null,
        __responseType: null,
        __originalResponseType: null,
    };
}
export function generateVideoStreamingUrl() {
    const payload = { fileId: ':fileId' };
    const getRequestOptions = ambassadorWixMediaSiteMediaV1FileDescriptor.generateVideoStreamingUrl(payload);
    const getUrl = (context) => {
        const { url } = getRequestOptions(context);
        return url;
    };
    return {
        getUrl,
        httpMethod: 'POST',
        path: '/v1/files/video/{fileId}/generate-stream-url',
        pathParams: { fileId: 'fileId' },
        __requestType: null,
        __originalRequestType: null,
        __responseType: null,
        __originalResponseType: null,
    };
}
export function bulkDeleteFiles() {
    const payload = {};
    const getRequestOptions = ambassadorWixMediaSiteMediaV1FileDescriptor.bulkDeleteFiles(payload);
    const getUrl = (context) => {
        const { url } = getRequestOptions(context);
        return url;
    };
    return {
        getUrl,
        httpMethod: 'POST',
        path: '/v1/bulk/files/delete',
        pathParams: {},
        __requestType: null,
        __originalRequestType: null,
        __responseType: null,
        __originalResponseType: null,
    };
}
export function bulkRestoreFilesFromTrashBin() {
    const payload = {};
    const getRequestOptions = ambassadorWixMediaSiteMediaV1FileDescriptor.bulkRestoreFilesFromTrashBin(payload);
    const getUrl = (context) => {
        const { url } = getRequestOptions(context);
        return url;
    };
    return {
        getUrl,
        httpMethod: 'POST',
        path: '/v1/bulk/trash-bin/files/restore',
        pathParams: {},
        __requestType: null,
        __originalRequestType: null,
        __responseType: null,
        __originalResponseType: null,
    };
}
export function listDeletedFiles() {
    const payload = {};
    const getRequestOptions = ambassadorWixMediaSiteMediaV1FileDescriptor.listDeletedFiles(payload);
    const getUrl = (context) => {
        const { url } = getRequestOptions(context);
        return url;
    };
    return {
        getUrl,
        httpMethod: 'GET',
        path: '/v1/trash-bin/files',
        pathParams: {},
        __requestType: null,
        __originalRequestType: null,
        __responseType: null,
        __originalResponseType: null,
    };
}
