// @wix/media/build/es/src/media-site-media-v1-file-descriptor.types.js
export var MediaType;
(function (MediaType) {
    MediaType["UNKNOWN"] = "UNKNOWN";
    MediaType["IMAGE"] = "IMAGE";
    MediaType["VIDEO"] = "VIDEO";
    MediaType["AUDIO"] = "AUDIO";
    MediaType["DOCUMENT"] = "DOCUMENT";
    MediaType["VECTOR"] = "VECTOR";
    MediaType["ARCHIVE"] = "ARCHIVE";
    MediaType["MODEL3D"] = "MODEL3D";
    MediaType["OTHER"] = "OTHER";
})(MediaType || (MediaType = {}));
export var OperationStatus;
(function (OperationStatus) {
    /** File upload or processing failed */
    OperationStatus["FAILED"] = "FAILED";
    /** File is ready for consumption */
    OperationStatus["READY"] = "READY";
    /** File is waiting for processing or currently being processed */
    OperationStatus["PENDING"] = "PENDING";
})(OperationStatus || (OperationStatus = {}));
export var State;
(function (State) {
    State["OK"] = "OK";
    State["DELETED"] = "DELETED";
})(State || (State = {}));
export var UploadProtocol;
(function (UploadProtocol) {
    /** The upload protocol to use for implementing the resumable upload. */
    UploadProtocol["TUS"] = "TUS";
})(UploadProtocol || (UploadProtocol = {}));
export var SortOrder;
(function (SortOrder) {
    SortOrder["ASC"] = "ASC";
    SortOrder["DESC"] = "DESC";
})(SortOrder || (SortOrder = {}));
export var RootFolder;
(function (RootFolder) {
    /** Root of all site media */
    RootFolder["MEDIA_ROOT"] = "MEDIA_ROOT";
    /** Root of the trash system folder */
    RootFolder["TRASH_ROOT"] = "TRASH_ROOT";
    /** Root of all visitor uploads */
    RootFolder["VISITOR_UPLOADS_ROOT"] = "VISITOR_UPLOADS_ROOT";
})(RootFolder || (RootFolder = {}));
export var StreamFormat;
(function (StreamFormat) {
    StreamFormat["UNKNOWN"] = "UNKNOWN";
    StreamFormat["HLS"] = "HLS";
    StreamFormat["DASH"] = "DASH";
})(StreamFormat || (StreamFormat = {}));
