// @wix/metro-runtime/dist/esm/serializer/velo-converters/video.js
var _a;
import querystring from 'querystring';
import { ConverterType } from '../domain';
import { alignIfLegacy, URL_HASH_PREFIX, WIX_PROTOCOL, } from './converters-utils';
var WIX_VIDEO = 'video';
export var video = (_a = {
        types: ['wix.common.Video']
    },
    _a[ConverterType.FROM_JSON] = {
        transform: function (val) {
            var _a;
            if (!val) {
                return;
            }
            var posterData = '';
            if ((_a = val.thumbnail) === null || _a === void 0 ? void 0 : _a.id) {
                posterData = "".concat(URL_HASH_PREFIX, "posterUri=").concat(val.thumbnail.id, "&posterWidth=").concat(val.thumbnail.width, "&posterHeight=").concat(val.thumbnail.height);
            }
            return val.id ? "wix:video://v1/".concat(val.id).concat(posterData) : val.url;
        },
    },
    _a[ConverterType.TO_JSON] = {
        transform: function (val) {
            if (!val) {
                return;
            }
            var alignedVideo = alignIfLegacy(val, WIX_VIDEO);
            var _a = new URL(alignedVideo), protocol = _a.protocol, hash = _a.hash, pathname = _a.pathname;
            var _b = querystring.parse(hash.replace(URL_HASH_PREFIX, '')), height = _b.posterHeight, width = _b.posterWidth, posterUri = _b.posterUri;
            var id = pathname.replace("".concat(WIX_VIDEO, "://v1/"), '').split('/')[0];
            if (protocol === WIX_PROTOCOL) {
                if (!posterUri) {
                    return { id: id };
                }
                return {
                    id: id,
                    thumbnail: {
                        height: Number(height),
                        width: Number(width),
                        id: posterUri.toString(),
                    },
                };
            }
            return { url: val };
        },
    },
    _a);
