// @wix/identity/build/es/src/iam-recovery-v1-recovery.http.js
import { serializer } from '@wix/metro-runtime/ambassador';
import { resolveUrl } from '@wix/metro-runtime';
const _commonCustomValue = {
    numValue: 'DOUBLE',
    dateValue: 'google.protobuf.Timestamp',
    listValue: '_commonListValue',
    mapValue: '_commonMapValue',
};
const _commonListValue = { value: '_commonCustomValue' };
const _commonMapValue = { value: 'Map#_commonCustomValue' };
const _customField = { value: '_customValue' };
const _customValue = {
    numValue: 'DOUBLE',
    dateValue: 'google.protobuf.Timestamp',
    listValue: '_listValue',
    mapValue: '_mapValue',
};
const _identity = {
    createdDate: 'google.protobuf.Timestamp',
    updatedDate: 'google.protobuf.Timestamp',
    identityProfile: '_identityProfile',
};
const _identityProfile = { customFields: '_customField' };
const _listValue = { value: '_customValue' };
const _mapValue = { value: 'Map#_customValue' };
const _recoverRequest = {};
const _sendActivationEmailRequest = {};
const _sendActivationEmailResponse = {};
const _sendRecoveryEmailRequest = {};
const _sendRecoveryEmailResponse = {};
const _stateMachineResponse = {
    identity: '_identity',
    additionalData: 'Map#_commonCustomValue',
};
function resolveWixIamRecoveryV1RecoveryServiceUrl(opts) {
    const domainToMappings = {
        _: [
            {
                srcPath: '/_iam/recovery',
                destPath: '',
            },
            {
                srcPath: '/_api/iam/recovery',
                destPath: '',
            },
        ],
        'www.wixapis.com': [
            {
                srcPath: '/_api/iam/recovery',
                destPath: '',
            },
        ],
    };
    return resolveUrl(Object.assign(opts, { domainToMappings }));
}
/**
 * Sends a member an email containing a customized link to a Wix-managed page
 * where the member can set a new password for their account.
 */
export function sendRecoveryEmail(payload) {
    const { toJSON: toReq, fromJSON: fromReq } = serializer(_sendRecoveryEmailRequest, {});
    const { fromJSON: fromRes } = serializer(_sendRecoveryEmailResponse, {});
    function __sendRecoveryEmail({ host }) {
        const serializedData = toReq(payload);
        const metadata = {
            entityFqdn: 'wix.iam.recovery.v1.recovery',
            method: 'POST',
            methodFqn: 'wix.iam.recovery.v1.RecoveryService.SendRecoveryEmail',
            url: resolveWixIamRecoveryV1RecoveryServiceUrl({
                protoPath: '/v1/send-email',
                data: serializedData,
                host,
            }),
            data: serializedData,
            transformResponse: fromRes,
        };
        return metadata;
    }
    __sendRecoveryEmail.fromReq = fromReq;
    return __sendRecoveryEmail;
}
/**
 * Sends an activation email with an activation token
 * making the transition from initial contact to a site member
 */
export function sendActivationEmail(payload) {
    const { toJSON: toReq, fromJSON: fromReq } = serializer(_sendActivationEmailRequest, {});
    const { fromJSON: fromRes } = serializer(_sendActivationEmailResponse, {});
    function __sendActivationEmail({ host }) {
        const serializedData = toReq(payload);
        const metadata = {
            entityFqdn: 'wix.iam.recovery.v1.recovery',
            method: 'POST',
            methodFqn: 'wix.iam.recovery.v1.RecoveryService.SendActivationEmail',
            url: resolveWixIamRecoveryV1RecoveryServiceUrl({
                protoPath: '/v1/activation-email',
                data: serializedData,
                host,
            }),
            data: serializedData,
            transformResponse: fromRes,
        };
        return metadata;
    }
    __sendActivationEmail.fromReq = fromReq;
    return __sendActivationEmail;
}
export function recover(payload) {
    const { toJSON: toReq, fromJSON: fromReq } = serializer(_recoverRequest, {});
    const { fromJSON: fromRes } = serializer(_stateMachineResponse, {
        _commonCustomValue,
        _commonListValue,
        _commonMapValue,
        _customField,
        _customValue,
        _identity,
        _identityProfile,
        _listValue,
        _mapValue,
    });
    function __recover({ host }) {
        const serializedData = toReq(payload);
        const metadata = {
            entityFqdn: 'wix.iam.recovery.v1.recovery',
            method: 'POST',
            methodFqn: 'wix.iam.recovery.v1.RecoveryService.Recover',
            url: resolveWixIamRecoveryV1RecoveryServiceUrl({
                protoPath: '/v1/recover',
                data: serializedData,
                host,
            }),
            data: serializedData,
            transformResponse: fromRes,
        };
        return metadata;
    }
    __recover.fromReq = fromReq;
    return __recover;
}
