// @wix/identity/build/es/src/identity-oauth-v1-refresh-token.http.js
import { toURLSearchParams } from '@wix/metro-runtime';
import { serializer } from '@wix/metro-runtime/ambassador';
import { resolveUrl } from '@wix/metro-runtime';
const _authorizeRequest = {};
const _deviceCodeRequest = {};
const _deviceCodeResponse = {};
const _deviceVerifyRequest = {};
const _deviceVerifyResponse = {};
const _rawHttpRequest = { body: 'BYTES' };
const _rawHttpResponse = { body: 'BYTES' };
const _revokeRefreshTokenRequest = {};
const _revokeRefreshTokenResponse = {};
function resolveWixIdentityOauth2V1Oauth2NgUrl(opts) {
    const domainToMappings = {
        'manage._base_domain_': [
            {
                srcPath: '/oauth2',
                destPath: '/v1/oauth',
            },
        ],
        'www.wixapis.com': [
            {
                srcPath: '/oauth2',
                destPath: '/v1/oauth',
            },
        ],
        'users._base_domain_': [
            {
                srcPath: '/v1/oauth/device/verify',
                destPath: '/v1/oauth/device/verify',
            },
            {
                srcPath: '/v1/oauth/manage/user-code',
                destPath: '/v1/oauth/manage/user-code',
            },
        ],
        _: [
            {
                srcPath: '/_api/oauth2',
                destPath: '/v1/oauth',
            },
        ],
    };
    return resolveUrl(Object.assign(opts, { domainToMappings }));
}
export function authorize(payload) {
    const { toJSON: toReq, fromJSON: fromReq } = serializer(_authorizeRequest, {});
    const { fromJSON: fromRes } = serializer(_rawHttpResponse, {});
    function __authorize({ host }) {
        const serializedData = toReq(payload);
        const metadata = {
            entityFqdn: 'wix.identity.oauth.v1.refresh_token',
            method: 'GET',
            methodFqn: 'wix.identity.oauth2.v1.Oauth2Ng.Authorize',
            url: resolveWixIdentityOauth2V1Oauth2NgUrl({
                protoPath: '/v1/oauth/authorize',
                data: serializedData,
                host,
            }),
            params: toURLSearchParams(serializedData),
            transformResponse: fromRes,
        };
        return metadata;
    }
    __authorize.fromReq = fromReq;
    return __authorize;
}
/**
 * Token endpoint accepts the TokenRequest message encoded in the raw http request body, and returns either
 * TokenResponse(status code = 200) or TokenErrorResponse (status code = 4XX) encoded in the raw http response body.
 */
export function token(payload) {
    const { toJSON: toReq, fromJSON: fromReq } = serializer(_rawHttpRequest, {});
    const { fromJSON: fromRes } = serializer(_rawHttpResponse, {});
    function __token({ host }) {
        const serializedData = toReq(payload);
        const metadata = {
            entityFqdn: 'wix.identity.oauth.v1.refresh_token',
            method: 'POST',
            methodFqn: 'wix.identity.oauth2.v1.Oauth2Ng.Token',
            url: resolveWixIdentityOauth2V1Oauth2NgUrl({
                protoPath: '/v1/oauth/token',
                data: serializedData,
                host,
            }),
            data: serializedData,
            transformResponse: fromRes,
        };
        return metadata;
    }
    __token.fromReq = fromReq;
    return __token;
}
/** this endpoint serves the Device Authorization Flow as described in https://www.rfc-editor.org/rfc/pdfrfc/rfc8628.txt.pdf */
export function deviceCode(payload) {
    const { toJSON: toReq, fromJSON: fromReq } = serializer(_deviceCodeRequest, {});
    const { fromJSON: fromRes } = serializer(_deviceCodeResponse, {});
    function __deviceCode({ host }) {
        const serializedData = toReq(payload);
        const metadata = {
            entityFqdn: 'wix.identity.oauth.v1.refresh_token',
            method: 'GET',
            methodFqn: 'wix.identity.oauth2.v1.Oauth2Ng.DeviceCode',
            url: resolveWixIdentityOauth2V1Oauth2NgUrl({
                protoPath: '/v1/oauth/device/code',
                data: serializedData,
                host,
            }),
            params: toURLSearchParams(serializedData),
            transformResponse: fromRes,
        };
        return metadata;
    }
    __deviceCode.fromReq = fromReq;
    return __deviceCode;
}
/** this endpoint serves the Device Authorization Flow as described in https://www.rfc-editor.org/rfc/pdfrfc/rfc8628.txt.pdf */
export function deviceVerify(payload) {
    const { toJSON: toReq, fromJSON: fromReq } = serializer(_deviceVerifyRequest, {});
    const { fromJSON: fromRes } = serializer(_deviceVerifyResponse, {});
    function __deviceVerify({ host }) {
        const serializedData = toReq(payload);
        const metadata = {
            entityFqdn: 'wix.identity.oauth.v1.refresh_token',
            method: 'GET',
            methodFqn: 'wix.identity.oauth2.v1.Oauth2Ng.DeviceVerify',
            url: resolveWixIdentityOauth2V1Oauth2NgUrl({
                protoPath: '/v1/oauth/device/verify',
                data: serializedData,
                host,
            }),
            params: toURLSearchParams(serializedData),
            transformResponse: fromRes,
        };
        return metadata;
    }
    __deviceVerify.fromReq = fromReq;
    return __deviceVerify;
}
/** Revoke a refresh token created by this service. */
export function revokeRefreshToken(payload) {
    const { toJSON: toReq, fromJSON: fromReq } = serializer(_revokeRefreshTokenRequest, {});
    const { fromJSON: fromRes } = serializer(_revokeRefreshTokenResponse, {});
    function __revokeRefreshToken({ host }) {
        const serializedData = toReq(payload);
        const metadata = {
            entityFqdn: 'wix.identity.oauth.v1.refresh_token',
            method: 'POST',
            methodFqn: 'wix.identity.oauth2.v1.Oauth2Ng.RevokeRefreshToken',
            url: resolveWixIdentityOauth2V1Oauth2NgUrl({
                protoPath: '/v1/oauth/revoke',
                data: serializedData,
                host,
            }),
            data: serializedData,
            transformResponse: fromRes,
        };
        return metadata;
    }
    __revokeRefreshToken.fromReq = fromReq;
    return __revokeRefreshToken;
}
