// @wix/motion-edm-autogen-p13n/dist/cjs/address/index.js
"use strict";

exports.__esModule = true;
exports.p13nToCorvid = exports.corvidToP13n = void 0;
var _lodash = require("lodash");
const sharedFieldNames = ['city', 'subdivision', 'country', 'postalCode'];
const sharedStreetAddressFieldNames = ['name', 'number', 'apt'];
const p13nToCorvid = p13nAddress => p13nAddress ? (0, _lodash.omitBy)({
  formatted: p13nAddress.formattedAddress,
  location: p13nAddress.geocode,
  ...(0, _lodash.pick)(p13nAddress, sharedFieldNames),
  streetAddress: p13nAddress.streetAddress && (0, _lodash.pick)(p13nAddress.streetAddress, sharedStreetAddressFieldNames),
  addressLine1: p13nAddress.addressLine,
  addressLine2: p13nAddress.addressLine2
}, _lodash.isUndefined) : undefined;
exports.p13nToCorvid = p13nToCorvid;
const corvidToP13n = corvidAddress => corvidAddress ? (0, _lodash.omitBy)({
  formattedAddress: corvidAddress.formatted,
  geocode: corvidAddress.location,
  ...(0, _lodash.pick)(corvidAddress, sharedFieldNames),
  streetAddress: corvidAddress.streetAddress && (0, _lodash.pick)(corvidAddress.streetAddress, sharedStreetAddressFieldNames),
  addressLine: corvidAddress.addressLine1,
  addressLine2: corvidAddress.addressLine2
}, _lodash.isUndefined) : undefined;
exports.corvidToP13n = corvidToP13n;
