// @wix/motion-edm-autogen-transformations/dist/esm/lib/getQueryItemsTransformationPaths.js
import { reduceTransformation, stripJsonPathRootPrefix, } from '@wix/motion-edm-autogen-transformations-core';
// TODO: Pass the item transformation instead of the entire response transformation
// in here. This will make this function simpler (we won't need to use reduceTransformation recursively)
var getQueryItemsTransformationPaths = function (transformation) {
    return reduceTransformation(transformation, {
        visitors: {
            MapArrayItems: function (accumulator, _parentPath, _sourceArrayExpression, itemTransformation) {
                return reduceTransformation(itemTransformation, {
                    visitors: {
                        NestedSimpleTransformationExpression: function (acc, path, value) {
                            acc[path] = stripJsonPathRootPrefix(value);
                        },
                    },
                    accumulator: accumulator,
                });
            },
        },
        accumulator: {},
    }) || {};
};
export { getQueryItemsTransformationPaths };
