// @wix/motion-edm-autogen-query-wrapper/dist/cjs/Iterator.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
exports.__esModule = true;
exports.Iterator = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
class Iterator {
  constructor({
    items,
    originQuery,
    fetchNextPage,
    fetchPrevPage,
    limit
  }) {
    (0, _defineProperty2.default)(this, "_items", void 0);
    (0, _defineProperty2.default)(this, "_fetchNextPage", void 0);
    (0, _defineProperty2.default)(this, "_fetchPrevPage", void 0);
    (0, _defineProperty2.default)(this, "_originQuery", void 0);
    (0, _defineProperty2.default)(this, "_limit", void 0);
    this._items = items;
    this._fetchNextPage = fetchNextPage;
    this._fetchPrevPage = fetchPrevPage;
    this._originQuery = originQuery;
    this._limit = limit;
  }
  get items() {
    return this._items;
  }
  get length() {
    return this._items.length;
  }
  get pageSize() {
    return this._limit;
  }
  get query() {
    return this._originQuery;
  }
  async next() {
    if (!this.hasNext()) {
      throw new Error('No next page to fetch');
    }
    const nextPageIterator = await this._fetchNextPage();
    return nextPageIterator;
  }
  async prev() {
    if (!this.hasPrev()) {
      throw new Error('No previous page to fetch');
    }
    const previousPageIterator = await this._fetchPrevPage();
    return previousPageIterator;
  }
}
exports.Iterator = Iterator;
