// @wix/motion-edm-autogen-p13n/dist/esm/image/index.js
import querystring from 'querystring';
import { toAltText } from './toAltText';
var WIX_IMAGE_PROTOCOL = 'wix:';
var WIX_LEGACY_IMAGE_PROTOCOL = 'image:';
var WIX_IMAGE_PROTOCOL_SUFFIX = 'image://v1/';
var URL_HASH_PREFIX = '#';
var DEFAULT_IMAGE_FILE_NAME = 'file';
var alignIfLegacyImage = function (image) {
    var protocol = new URL(image).protocol;
    return protocol === WIX_LEGACY_IMAGE_PROTOCOL ? "wix:".concat(image) : image;
};
var corvidToP13n = function (image) {
    if (!image) {
        return undefined;
    }
    var alignedImage = alignIfLegacyImage(image);
    var _a = new URL(alignedImage), protocol = _a.protocol, hash = _a.hash, pathname = _a.pathname;
    var _b = querystring.parse(hash.replace(URL_HASH_PREFIX, '')), height = _b.originHeight, width = _b.originWidth;
    var _c = pathname
        .replace(WIX_IMAGE_PROTOCOL_SUFFIX, '')
        .split('/'), id = _c[0], filename = _c[1];
    if (protocol === WIX_IMAGE_PROTOCOL) {
        return {
            id: id,
            height: Number(height),
            width: Number(width),
            altText: toAltText(filename),
        };
    }
    return { url: image };
};
var computeImageFileName = function (image) {
    var _a;
    var filename = (_a = image.altText) !== null && _a !== void 0 ? _a : DEFAULT_IMAGE_FILE_NAME;
    return encodeURIComponent(filename);
};
var p13nToCorvid = function (image) {
    if (!image) {
        return undefined;
    }
    return image.id
        ? "wix:image://v1/".concat(image.id, "/").concat(computeImageFileName(image), "#originWidth=").concat(image.width, "&originHeight=").concat(image.height)
        : image === null || image === void 0 ? void 0 : image.url;
};
export { corvidToP13n, p13nToCorvid };
