// @wix/redirects/build/cjs/src/headless-v1-redirect-session.types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CallbackType = exports.MembersAccountSection = exports.Prompt = exports.LocationType = void 0;
var LocationType;
(function (LocationType) {
    LocationType["UNDEFINED"] = "UNDEFINED";
    LocationType["OWNER_BUSINESS"] = "OWNER_BUSINESS";
    LocationType["OWNER_CUSTOM"] = "OWNER_CUSTOM";
    LocationType["CUSTOM"] = "CUSTOM";
})(LocationType = exports.LocationType || (exports.LocationType = {}));
/** Currently only `none` and `login` are supported. */
var Prompt;
(function (Prompt) {
    Prompt["login"] = "login";
    Prompt["none"] = "none";
    Prompt["consent"] = "consent";
    Prompt["select_account"] = "select_account";
})(Prompt = exports.Prompt || (exports.Prompt = {}));
var MembersAccountSection;
(function (MembersAccountSection) {
    /** default - account info section in "my account" */
    MembersAccountSection["ACCOUNT_INFO"] = "ACCOUNT_INFO";
    /** My Bookings section in "my account" */
    MembersAccountSection["BOOKINGS"] = "BOOKINGS";
    /** My Orders section in "my account" */
    MembersAccountSection["ORDERS"] = "ORDERS";
    /** Subscriptions section in "my account" */
    MembersAccountSection["SUBSCRIPTIONS"] = "SUBSCRIPTIONS";
    /** Events section in "my account" */
    MembersAccountSection["EVENTS"] = "EVENTS";
})(MembersAccountSection = exports.MembersAccountSection || (exports.MembersAccountSection = {}));
var CallbackType;
(function (CallbackType) {
    /** Invalid value - default */
    CallbackType["UNKNOWN"] = "UNKNOWN";
    /** The callback URL is used for the logout flow */
    CallbackType["LOGOUT"] = "LOGOUT";
    /** The callback URL is used for a checkout flow */
    CallbackType["CHECKOUT"] = "CHECKOUT";
    /** The callback URL is used for the authorize flow */
    CallbackType["AUTHORIZE"] = "AUTHORIZE";
})(CallbackType = exports.CallbackType || (exports.CallbackType = {}));
