// @wix/redirects/build/cjs/src/headless-v1-redirect-session.http.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRedirectSession = void 0;
const ambassador_1 = require("@wix/metro-runtime/ambassador");
const metro_runtime_1 = require("@wix/metro-runtime");
const _createRedirectSessionRequest = {};
const _createRedirectSessionResponse = {};
function resolveWixHeadlessV1RedirectSessionServiceUrl(opts) {
    const domainToMappings = {
        'www._base_domain_': [
            {
                srcPath: '/_api/redirects-api',
                destPath: '',
            },
        ],
        'www.wixapis.com': [
            {
                srcPath: '/_api/redirects-api',
                destPath: '',
            },
            {
                srcPath: '/redirect-session',
                destPath: '',
            },
        ],
    };
    return (0, metro_runtime_1.resolveUrl)(Object.assign(opts, { domainToMappings }));
}
/**
 * Creates a URL for redirecting a visitor from an external client site to a Wix page for Wix-managed functionality.
 *
 *
 * The `createRedirectSession()` function enables your external Wix Headless client site, built on any platform, to integrate Wix-managed frontend functionality for specific processes.
 * For example, your site can temporarily redirect a visitor to Wix for authentication, or for a checkout process for a bookings, eCommerce, events, or paid plans transaction.
 *
 * To initiate a redirect session:
 *
 * 1. Call `createRedirectSession()` with the details required for Wix to take care of one specific process (for example, authentication or a bookings checkout). Provide one or more callback URLs, so Wix can redirect the user back to your site as appropriate when the process is over.
 * 1. Redirect your visitor to the URL provided in the response. This URL includes query parameters informing Wix where to redirect the visitor back to on your external site.
 * 1. Make sure the pages at the callback URLs you provided take care of the next stages in your visitor flow.
 */
function createRedirectSession(payload) {
    const { toJSON: toReq, fromJSON: fromReq } = (0, ambassador_1.serializer)(_createRedirectSessionRequest, {});
    const { fromJSON: fromRes } = (0, ambassador_1.serializer)(_createRedirectSessionResponse, {});
    function __createRedirectSession({ host }) {
        const serializedData = toReq(payload);
        const metadata = {
            entityFqdn: 'wix.headless.v1.redirect_session',
            method: 'POST',
            methodFqn: 'wix.headless.v1.RedirectSessionService.CreateRedirectSession',
            url: resolveWixHeadlessV1RedirectSessionServiceUrl({
                protoPath: '/v1/redirect-session',
                data: serializedData,
                host,
            }),
            data: serializedData,
            transformResponse: fromRes,
        };
        return metadata;
    }
    __createRedirectSession.fromReq = fromReq;
    return __createRedirectSession;
}
exports.createRedirectSession = createRedirectSession;
