// @wix/motion-edm-autogen-transformations/dist/cjs/lib/getQueryItemsTransformationPaths.js
"use strict";

exports.__esModule = true;
exports.getQueryItemsTransformationPaths = void 0;
var _motionEdmAutogenTransformationsCore = require("@wix/motion-edm-autogen-transformations-core");
// TODO: Pass the item transformation instead of the entire response transformation
// in here. This will make this function simpler (we won't need to use reduceTransformation recursively)
const getQueryItemsTransformationPaths = transformation => (0, _motionEdmAutogenTransformationsCore.reduceTransformation)(transformation, {
  visitors: {
    MapArrayItems: (accumulator, _parentPath, _sourceArrayExpression, itemTransformation) => (0, _motionEdmAutogenTransformationsCore.reduceTransformation)(itemTransformation, {
      visitors: {
        NestedSimpleTransformationExpression(acc, path, value) {
          acc[path] = (0, _motionEdmAutogenTransformationsCore.stripJsonPathRootPrefix)(value);
        }
      },
      accumulator
    })
  },
  accumulator: {}
}) || {};
exports.getQueryItemsTransformationPaths = getQueryItemsTransformationPaths;
